File: 3. Longest Substring Without Repeating Characters Medium/solution.py
def longest_substring_length(s: str) -> int:
    last_occurrence = [-1] * 256
    start = 0
    max_length = 0

    for i, char in enumerate(s):
        char_code = ord(char)
        if last_occurrence[char_code] >= start:
            start = last_occurrence[char_code] + 1
        last_occurrence[char_code] = i
        max_length = max(max_length, i - start + 1)

    return max_length
        

print(longest_substring_length("abcabcbb"))
print(longest_substring_length("bbbbb"))
print(longest_substring_length("pwwkew"))

File: 3. Longest Substring Without Repeating Characters Medium/solution.java
import java.util.Arrays;

public class solution {
    public static int longestSubstringLength(String s) {
        int[] lastOccurrence = new int[256];
        Arrays.fill(lastOccurrence, -1);
        int start = 0;
        int maxLength = 0;

        for (int i = 0; i < s.length(); i++) {
            char currentChar = s.charAt(i);
            int charCode = (int) currentChar;
            if (lastOccurrence[charCode] >= start) {
                start = lastOccurrence[charCode] + 1;
            }
            lastOccurrence[charCode] = i;
            maxLength = Math.max(maxLength, i - start + 1);
        }

        return maxLength;
    }

    public static void main(String[] args) {
        System.out.println(longestSubstringLength("abcabcbb"));
        System.out.println(longestSubstringLength("bbbbb"));
        System.out.println(longestSubstringLength("pwwkew"));
    }
}
